### DISCUSSION TITLE
RFC: Improve relation between orderGrantedRefund and TransactionItem

### DISCUSSION THREAD

**Discussion by korycins** [saleor/saleor#15458] [RFC]
### Problem

Currently, we don't have any relationship between `OrderGrantedRefund` and `TransactionItem` related to the refund. 
Because of this, making a relation on the dashboard side (for displaying purposes) grantedRefund+Transaction is impossible. 
The RFC proposed the changes in mentioned types, to make the relation between grantedRefund and Transaction when staff user creates the grantedRefund

### 1. General Assumptions

- When creating new OrderGrantedRefund, allow assigning related TransactionItem (in future versions this will be mandatory step)
- When updating orderGrantedRefund, allow changing the assigned TransactionItem  object.
- The amount assigned to granted refund, cannot be greater than `transactionItem.chargedAmount`
- In case when transactionItem's chargedAmount will be reduced by a received request from payment app, the `TransactionRequestRefundForGrantedRefund` will raise an exception
- `TransactionRequestRefundForGrantedRefund` mutation will not require `transactionItem` as ID, when transaction is assigned to orderGrantedRefund
- OrderGrantedRefund will have a status field, transactionItem, and list of `transactionEvents` related to the refund actions
- Retry will be possible by calling a second-time `TransactionRequestRefundForGrantedRefund`, when the previous refund action was marked as a fail.
- order granted refund status should be calculated based on the transactionEvents assigned to grantedRefund.
- The transactionEvent REFUND_REQUEST created when calling `TransactionRequestRefundForGrantedRefund` should be connected with grantedRefund
- When new event with type REFUND_SUCCESS/REFUND_FAILURE is created by transactionEventReport mutation, it should be connected to granted refund (if psp-reference matches to REFUND_REQUEST event, created when calling `TransactionRequestRefundForGrantedRefund`)
- orderGrantRefundUpdate should be blocked when status is FULL or PENDING. Only reason can be updated in that case.

### Mutations

```graphql
input OrderGrantRefundCreateInput @doc(category: "Orders") {
  ...
  """
  Transaction ID related to the granted refund. transaction.chargedAmount need to be
  equal or greater than provided `amount`. If `amount` calculated automatically, the
  min(amount, transaction.chargedAmount) will be used.
  Field will be required starting from Saleor 3.20.
  """
  transactionId: ID
}

input OrderGrantRefundUpdateInput @doc(category: "Orders") {
  ...
  """
  Transaction ID related to the granted refund. transaction.chargedAmount needs to be
  equal or greater than provided `amount`. If `amount` calculated automatically, the
  min(amount, transaction.chargedAmount) will be used.
  """
  transactionId: ID
}

transactionRequestRefundForGrantedRefund(
    """The ID of the granted refund."""
    grantedRefundId: ID!

    """The ID of the transaction. One of field: id or token is required, when transaction is not assigned to grantedRefund."""
    id: ID

    """
    The token of the transaction. One of field:id or token is required, when transaction is not assigned to grantedRefund.
    
    Added in Saleor 3.14.
    """
    token: UUID
  ): TransactionRequestRefundForGrantedRefund @doc(category: "Payments")
```

### Types

```graphql
enum OrderGrantedRefundStatusEnum @doc(category: "Orders") {
  NONE
  PENDING
  FULL
  FAIL
}

type OrderGrantedRefund @doc(category: "Orders") {
    """
    Represents the current status of the granted refund
    """
    status: OrderGrantedRefundStatusEnum!
    """
    The transactionItem assigned to the granted refund
    """
    transaction: TransactionItem
    """
    List of refund events related to the granted refund.
    """
    transactionEvents: [TransactionEvent!]!
}
```

### 3. Database changes

`order/models.py`
```python
class OrderGrantedRefund:
    status = models.CharField(
        choices=OrderGrantedRefundStatus.CHOICES
    )
    transaction = models.OneToOneField(
        TransactionItem, related_name="granted_refund", on_delete=models.SET_NULL()
    )
```
`payment/models.py`
```python
class TransactionEvent(models.Model):
      related_granted_refund = models.ForeignKey(
        "order.OrderGrantedRefund",
        related_name="transaction_events",
        on_delete=models.SET_NULL,
    )
```

### 5. To Do list

- [ ] docs are up-to-date
- [ ] order granted refund mutations have updated inputs
- [ ] order granted refund mutations have validation for amount vs transactionItem.chargedAmount
- [ ] new fields are added to orderGrantedRefund type (status, transaction, transactionEvents)
- [ ] related refund events  are assigned to orderGrantedEvent and listed on the orderGrantedRefund type
- [ ] status of the orderGrantedRefund is calculated based on the assigned events
- [ ] TransactionRequestRefundForGrantedRefund has proper validation, to confirm that refund can be called on transaction assigned to grantedRefund (grantedRefund.amount <= transaction.chargedAmount)

### 6. To Test

TBD

**Comment by yellowee** [comment]
Reviewed by QA ✅
